### PR TITLE
Dask friendly check in `.weighted()`

### LINKED ISSUE #4541: Option to skip tests in `weighted()`
When working with large dask-array weights, [this check](https://github.com/pydata/xarray/blob/adc55ac4d2883e0c6647f3983c3322ca2c690514/xarray/core/weighted.py#L103-L107) triggers computation of the array. This affects [xgcms ability to layer operations lazily](https://github.com/xgcm/xgcm/blob/f0a0afd666184dd7d15bcff6f22ac50716b6c78a/xgcm/grid.py#L1737)

Would you be open to implement an option to skip this test, maybe with a warning displayed? Happy to submit a PR.

### PR DESCRIPTION
**Pull request by jbusecke**
- [x] Closes #4541 
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

File: xarray/core/weighted.py
Comment (max-sixty) [inline xarray/core/weighted.py:110]:
```suggestion
            return w
```

**Review by max-sixty — COMMENTED** [review]
Looks good! Thanks @jbusecke . I'll let @dcherian or @mathause  comment too.

**Comment by jbusecke** [comment]
The Ci environments without dask are failing. Should I add some pytest skip logic, or what is the best way to handle this?

**Comment by max-sixty** [comment]
Yes, `@requires_dask` on the test

**Comment by keewis** [comment]
since the test you added requires `dask` it definitely should get a `@requires_dask`.

That won't fix all the failing tests, though: `map_blocks` is only valid if we operate on `dask` arrays. I guess you either have to apply the function directly on other arrays, or we have to make `DataArray.map_blocks` on non-`dask` arrays call to `DataArray.map`.

File: xarray/tests/test_weighted.py
Comment (mathause) [inline xarray/tests/test_weighted.py:42]: You test that dask does not comoute: 

https://github.com/pydata/xarray/blob/83884a1c6dac4b5f6309dfea530414facc100bc8/xarray/tests/test_dask.py#L189-L190

**Review by mathause — COMMENTED** [review]
For simplicity I would use `if dask_duck_array(weights):`.

File: xarray/core/weighted.py
Comment (dcherian) [inline xarray/core/weighted.py:112]:
```suggestion
        weights = dask.array.map_blocks(_weight_check.data)
```

should be more efficient.

Comment (jbusecke) [reply xarray/core/weighted.py]: This seems to cause a bunch of problems for me. Could you explain why this is more efficient? I might be missing something due to my lack of experience with these internals.

Comment (mathause) [reply xarray/core/weighted.py:112]: `weights.map_blocks` wraps `dask.array.map_blocks` and does a bunch of extra stuff. So calling `dask.array.map_blocks` is more efficient.

**Comment by mathause** [comment]
I think you need to do someting along the lines of:

```python
if dask_duck_array(weights):
    import dask.array as dsa
    dsa.map_blocks(_weight_check, weights.data, dtype=weights.dtype)
else:
    _weight_check()
```

**Comment by jbusecke** [comment]
I did have to fiddle with this a bit. I did change `.isnull()` to `np.isnan()` and replace the data of `weights`,  otherwise the error would not be raised for the dask array at all. 
This does not look terribly elegant to me, but passes tests locally. Waiting for the CI to see if the others pass aswell. Happy to make further changes, and thanks for all the input already.

File: xarray/core/weighted.py
Comment (keewis) [inline xarray/core/weighted.py:107]: `.isnull()` does a bit more than that: `np.isnan` won't detect `NaT`. @mathause, how likely is it to get datetime-like arrays here? They don't make much sense as weights, but as far as I can tell we don't check (I might be missing something, though)

Comment (mathause) [reply xarray/core/weighted.py:107]: There is no check for that. A `TimeDelta` may make some sense as weights. `DateTime` not so much. I think we can get away with using `np.isnan`. A `Date*` array as weights containing `NaT` should be super uncommon.

Comment (jbusecke) [reply xarray/core/weighted.py:107]: We could still operate on the dataarray instead of the dask/numpy array, but as @dcherian suggesred, that would be less efficient. I would be curious as to what penalties would actually occur when we use the `weights.map_blocks` compared to `dask.array.map_blocks`?

**Comment by jbusecke** [comment]
Do you think this works or are further changes needed? Many thanks for the guidance so far!

File: xarray/core/weighted.py
Comment (max-sixty) [inline xarray/core/weighted.py:107]:
```suggestion
        def _weight_check(w):
            # Ref https://github.com/pydata/xarray/pull/4559/files#r515968670
            if np.isnan(w).any():
```

**Review by max-sixty — APPROVED** [review]
Looks great to me. Want to add a whatsnew?

File: xarray/core/weighted.py
Comment (dcherian) [reply xarray/core/weighted.py:107]: 1. Could use `duck_array_ops.isnull` to account for `timedelta64`? It is weird to have it as a weight though. Does that work?
2. Re map_blocks: the xarray version adds tasks that create xarray objects wrapping every block in a dask array. That adds overhead which is totally unneccesary here.

File: xarray/core/weighted.py
Comment (dcherian) [inline xarray/core/weighted.py:117]:
```suggestion
            weights = weights.copy(data=dsa.map_blocks(
```
so we don't modify the original object.

**Comment by jbusecke** [comment]
I am getting some failures for `py38-upstream` (TestDataset.test_polyfit_warnings) that seem unrelated?

**Comment by jbusecke** [comment]
Similarly on `MacOSX py38` this fails: TestDask.test_save_mfdataset_compute_false_roundtrip

**Comment by keewis** [comment]
the `TestDask.test_save_mfdataset_compute_false_roundtrip` failure should be #4539, and the failing `upstream-dev` CI shouldn't be related, either (my guess is that a dependency – most likely `numpy` – added a new warning). I'll investigate and open a new issue for that.

**Comment by jbusecke** [comment]
Ok I think this should be good to go. I have implemented all the requested changes. The remaining failures are related to other problems upstream (I think). Anything else I should add here?

File: xarray/core/weighted.py
Comment (dcherian) [inline xarray/core/weighted.py:120]:
```suggestion
            weights = weights.copy(data=_weight_check(weights.data))
```

File: xarray/core/weighted.py
Comment (mathause) [inline xarray/core/weighted.py:120]:
```suggestion
           _weight_check(weights.data)
```
Do we actually need to assign?  Is the code not executed?

**Comment by jbusecke** [comment]
I am not understanding why that `MinimumVersionPolicy` test is failing (or what it does haha). Is this something upstream, or should I fix?

**Comment by keewis** [comment]
this CI is sometimes flaky, it's usually enough to rerun it. I'll do that for you once the other CI finished.

**Comment by jbusecke** [comment]
Seems like all the other test are passing (minus the two upstream problems discussed before).

**Review by dcherian — APPROVED** [review]
Thanks @jbusecke great contribution!

File: xarray/core/weighted.py
Comment (mathause) [inline xarray/core/weighted.py:117]:
```suggestion
            weights.data.map_blocks(_weight_check, dtype=weights.dtype)
```

Comment (jbusecke) [reply xarray/core/weighted.py]: This wont trigger the check for dask input. I tried that before. It never gets computed.

File: xarray/core/weighted.py
Comment (mathause) [inline xarray/core/weighted.py:115]:
```suggestion
            # assign to copy - else the check is not triggered
            weights = weights.copy(
                data=weights.data.map_blocks(_weight_check, dtype=weights.dtype)
            )
```

File: xarray/core/weighted.py
Comment (mathause) [reply xarray/core/weighted.py:118]: Thanks for the explanation! I reverted the change & added a comment.

**Comment by mathause** [comment]
Thanks @jbusecke

**Pull request merged**
